eth/tracers/native/statediff.go

```diff
@@ -0,0 +1,190 @@
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+package native
+
+import (
+	"encoding/json"
+	"math/big"
+	"sync/atomic"
+	"time"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/common/hexutil"
+	"github.com/ethereum/go-ethereum/core/vm"
+	"github.com/ethereum/go-ethereum/crypto"
+	"github.com/ethereum/go-ethereum/eth/tracers"
+)
+
+func init() {
+	register("stateDiffTracer", newStateDiffTracer)
+}
+
+type diffstate = map[common.Address]*diffaccount
+type diffaccount struct {
+	Before account `json:"before"`
+	After  account `json:"after"`
+}
+
+type stateDiffTracer struct {
+	env       *vm.EVM
+	diffstate diffstate
+	create    bool
+	to        common.Address
+	gasLimit  uint64 // Amount of gas bought for the whole tx
+	interrupt uint32 // Atomic flag to signal execution interruption
+	reason    error  // Textual reason for the interruption
+}
+
+func newStateDiffTracer(ctx *tracers.Context) tracers.Tracer {
+	// First callframe contains tx context info
+	// and is populated on start and end.
+	return &stateDiffTracer{diffstate: diffstate{}}
+}
+
+// CaptureStart implements the EVMLogger interface to initialize the tracing operation.
+func (t *stateDiffTracer) CaptureStart(env *vm.EVM, from common.Address, to common.Address, create bool, input []byte, gas uint64, value *big.Int) {
+	t.env = env
+	t.create = create
+	t.to = to
+
+	t.lookupAccount(from)
+	t.lookupAccount(to)
+
+	// The recipient balance includes the value transferred.
+	toBal := hexutil.MustDecodeBig(t.diffstate[to].Before.Balance)
+	toBal = new(big.Int).Sub(toBal, value)
+	t.diffstate[to].Before.Balance = hexutil.EncodeBig(toBal)
+
+	// The sender balance is after reducing: value and gasLimit.
+	// We need to re-add them to get the pre-tx balance.
+	fromBal := hexutil.MustDecodeBig(t.diffstate[from].Before.Balance)
+	gasPrice := env.TxContext.GasPrice
+	consumedGas := new(big.Int).Mul(gasPrice, new(big.Int).SetUint64(t.gasLimit))
+	fromBal.Add(fromBal, new(big.Int).Add(value, consumedGas))
+	t.diffstate[from].Before.Balance = hexutil.EncodeBig(fromBal)
+	t.diffstate[from].Before.Nonce--
+}
+
+// CaptureEnd is called after the call finishes to finalize the tracing.
+func (t *stateDiffTracer) CaptureEnd(output []byte, gasUsed uint64, _ time.Duration, err error) {
+	if t.create {
+		// Exclude created contract.
+		delete(t.diffstate, t.to)
+	}
+
+	for addr, diff := range t.diffstate {
+		for key := range diff.Before.Storage {
+			t.diffstate[addr].After.Storage[key] = t.env.StateDB.GetState(addr, key)
```

Comment:
Please do this in `CaptureTxEnd`. Then the `after` state will also reflect the refunds and tip given to coinbase.

Reply:
thanks for your advice.